### PR TITLE
Resolve softmax issues with pytorch_model.py

### PR DESCRIPTION
**Pull request by dilyabareeva**
### Description
- Extraction of softmax layer from pytorch models sometimes lead to problem.
- We want to update the logic of softmax adjustment as follows:
IMG_4F0CB8FDD9CB-1-1

### Implemented changes
  - [x] Updated logic for softmax adjustments in the model depending on the model structure
  - [x] The 6 different cases from the photo represented in get_softmax_arg_model method (see comments)
  - [x] A better way to extract Softmax layer from model by replacing it with Identity (instead of rebuilding the model and forward function)

### CODE REVIEW

**Review by annahedstroem — CHANGES_REQUESTED** [review]
A few things! Thanks @dilyabareeva !

File: quantus/helpers/model/pytorch_model.py [hunk 1]
```diff
@@ -54,27 +54,36 @@ def __init__(
...
-        we have to rebuild the model and replace top with linear activation.
-        Cache the rebuilt model and reuse it during consecutive predict calls.
+        In a case model has a softmax module, the last torch.nn.Softmax module in the self.model.modules() list is
+        replaced with torch.nn.Identity().
         """
-        if not self._last_layer_is_softmax:
-            return self.model
+        linear_model = copy.deepcopy(self.model)
+
+        for attr_tpl in list(linear_model.named_modules())[::-1]:
```
Comment (aaarrti) [inline quantus/helpers/model/pytorch_model.py:81]: What does `attr_tpl` stand for? 
Please either use a more verbose self-descriptive variable name, or leave a comment on what does it mean.

**Comment by aaarrti** [comment]
We are modifying the structure of the model provided by the user, it would be great to make it visible to the user what exactly is being done under the hood, and what assumptions were made. 
E.g., for the case when the last softmax layer is replaced by a linear one, we could log a message: 
```
<XAI method name> reuires model with linear top, but found softmax activation on top for <model name>. Removed <layer name> at <layer index>
```

This must not be a warning, since this is expected behaviour. I'd suggest using `logging.info`, and set log format to something like `"[%(filename)s:%(lineno)s->%(funcName)s()]:%(levelname)s: %(message)s"`, so the user can see where it is coming from.

File: quantus/helpers/model/pytorch_model.py [hunk 2]
```diff
@@ -5,15 +5,15 @@
 # Quantus is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more details.
```
[the remark this answers is not in the document]
Comment (dilyabareeva) [reply quantus/helpers/model/pytorch_model.py:5]: Both use the same ModelInterface - so the public methods are the same. The logic in tf_model might need to be changed similarly: I'm not 100% sure that model.layers[-1] reliably returns the actual output layer in tensorflow. But this PR is about PyTorch, so maybe this is worth opening another issue to further investigate.

File: quantus/helpers/model/pytorch_model.py [hunk 3]
```diff
@@ -54,27 +54,36 @@ def __init__(
...
+        modules = list(self.model.modules())
+        for i in range(-1, -len(modules), -1):
+            if isinstance(modules[i], torch.nn.Softmax):
+                return i
+        return None
+
         last_layer = list(self.model.children())[-1]
         return isinstance(last_layer, torch.nn.Softmax)
 
-    @cachedmethod(operator.attrgetter("cache"))
```
[the remark this answers is not in the document]
Comment (dilyabareeva) [reply quantus/helpers/model/pytorch_model.py:67]: We changed the method in a way that we create a deepcopy of a model before we replace Softmax with Identity (we don't want to meddle with the original model of the user, and only want to create a non-softmax copy for metric calculations). The cachedmethod attributor seems to have problems with persisting very large objects like pytorch model, but the lru_cache seems to be working fine. See also https://docs.python.org/3/faq/programming.html#faq-cache-method-calls.

**Review by annahedstroem — COMMENTED** [review]
OK.
